src/volume/volume_multi.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+	Lightmetrica - Copyright (c) 2019 Hisanari Otsu
+	Distributed under MIT license. See LICENSE file for details.
+*/
+
+#include <pch.h>
+#include <lm/core.h>
+#include <lm/volume.h>
+
+LM_NAMESPACE_BEGIN(LM_NAMESPACE)
+
+/*
+\rst
+.. function:: volume::multigaussian
+
+    Multiple Gaussian volumes.
+
+    :param volumes_alb: Array of references to volume albedos
+    :param volumes_den: Array of references to volume densities
+\endrst
+*/
+class Volume_Multi : public Volume {
+private:
+    Bound bound_; // Boundary including all volumes
+    std::vector< Volume* > volumes_den_;
+    std::vector< Volume* > volumes_alb_;
+    unsigned int size_; // Size of volume arrays
+    Float maxScalar_ = 0; // Sum of maxScalar() of all Volumes in volumes_den_
+    Rng* rng; // Currently no way to get current thread's rng, fallback to this
```

Comment:
Please use `Rng` instead of `Rng*`. Current code didn't have delete of the pointer (memory leak), but you don't need if you use `Rng`.

Reply:
Rng.u() is not const, and eval_color is declared const, using Rng without a pointer makes mutable needed, which, I think, is not better.

Reply:
That's right. Also I noticed `eval_color` can be accessed from multiple threads and `Rng` is not thread-safe. I think it's better to avoid to use `Rng` in this context.  An workaround is to directly evaluate expected values of the current estimate, that is, taking a sum of scalar(i) * color(i) / C for volume index i, where C is normalization constant (= \sum_j scalar(j)).